autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/FreeIpaCommunicator.java

```diff
@@ -33,17 +42,26 @@ public FreeIpaCommunicator(UserV1Endpoint userV1Endpoint) {
     public SyncOperationStatus synchronizeAllUsers(SynchronizeAllUsersRequest request) {
         LOGGER.info("Invoking freeIpa user sync request: {}", request);
         String envCrn = request.getEnvironments().iterator().next();
-        SyncOperationStatus lastSyncStatus = ThreadBasedUserCrnProvider.doAsInternalActor(
-                internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
-                () -> userV1Endpoint.getLastSyncOperationStatus(envCrn));
-        if (SynchronizationStatus.RUNNING.equals(lastSyncStatus.getStatus())) {
-            LOGGER.info("There is a user sync operation already running for environment: {} with operationId: {}, " +
-                    "skipping request to trigger another user sync", envCrn, lastSyncStatus.getOperationId());
-            return lastSyncStatus;
-        } else {
-            return ThreadBasedUserCrnProvider.doAsInternalActor(
+        try {
+            SyncOperationStatus lastSyncStatus = ThreadBasedUserCrnProvider.doAsInternalActor(
                     internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
-                    () -> userV1Endpoint.synchronizeAllUsers(request));
+                    () -> userV1Endpoint.getLastSyncOperationStatus(envCrn));
+            if (SynchronizationStatus.RUNNING.equals(lastSyncStatus.getStatus())) {
+                LOGGER.info("There is a user sync operation already running for environment: {} with operationId: {}, " +
+                        "skipping request to trigger another user sync", envCrn, lastSyncStatus.getOperationId());
+                return lastSyncStatus;
+            } else {
+                SyncOperationStatus status = ThreadBasedUserCrnProvider.doAsInternalActor(
+                        internalCrnGeneratorFactory.autoscale().getInternalCrnForServiceAsString(),
+                        () -> userV1Endpoint.synchronizeAllUsers(request));
+                metricService.incrementMetricCounter(IPA_USER_SYNC_INVOCATION, MetricTag.TENANT.name(), request.getAccountId());
+                return status;
+            }
+        } catch (Exception ex) {
+            LOGGER.error("Encountered error when trying to synchronize users to IPA for environment: {}", envCrn, ex);
+            metricService.incrementMetricCounter(IPA_USER_SYNC_FAILED, MetricTag.TENANT.name(), request.getAccountId());
```

Comment:
If the SynchronizationStatus is REJECTED than should it be considered in the failed one as in this case the sync process might not have been performed.

Reply:
That is already being done because it looks like the REJECTED state is set if a SyncOperationAlreadyRunningException is thrown. The exception is then handled to increment failure count.

https://github.com/hortonworks/cloudbreak/blob/25f11bef48b7b9b2cb7985da70f6538b734b122a/freeipa/src/main/java/com/sequenceiq/freeipa/controller/UserV1Controller.java#L202-L207

Reply:
This looks ok to me as is. The API itself is not very well defined in terms of what SyncStatus can be returned. Beyond REJECTED, there's other states like FAILED. Do we need to be checking for these? (The API is not very clear on this)